runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsTranslation.java

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.metrics;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.runners.core.construction.metrics.MetricKey;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.metrics.MetricName;
+
+/** Translation utilities for metrics classes to/from Fn API. */
+@Experimental(Kind.METRICS)
+public abstract class MetricsTranslation {
+
+  private MetricsTranslation() {}
+
+  public static MetricUpdates metricUpdatesFromProto(
+      String ptransformName, Collection<BeamFnApi.Metrics.User> userMetricUpdates) {
+    List<MetricUpdates.MetricUpdate<Long>> counterUpdates = new ArrayList<>();
+    List<MetricUpdates.MetricUpdate<DistributionData>> distributionUpdates = new ArrayList<>();
+    List<MetricUpdates.MetricUpdate<GaugeData>> gaugeUpdates = Collections.emptyList();
+
+    for (BeamFnApi.Metrics.User userMetricUpdate : userMetricUpdates) {
+      MetricKey metricKey =
+          MetricKey.create(ptransformName, metricNameFromProto(userMetricUpdate.getMetricName()));
+      switch (userMetricUpdate.getDataCase()) {
+        case COUNTER_DATA:
+          counterUpdates.add(
+              MetricUpdates.MetricUpdate.create(
+                  metricKey, userMetricUpdate.getCounterData().getValue()));
+          break;
+        case DISTRIBUTION_DATA:
+          distributionUpdates.add(
+              MetricUpdates.MetricUpdate.create(
+                  metricKey,
+                  DistributionData.create(
+                      userMetricUpdate.getDistributionData().getSum(),
+                      userMetricUpdate.getDistributionData().getCount(),
+                      userMetricUpdate.getDistributionData().getMin(),
+                      userMetricUpdate.getDistributionData().getMax())));
+          break;
+        case DATA_NOT_SET:
+          continue;
+      }
+    }
+    return MetricUpdates.create(counterUpdates, distributionUpdates, gaugeUpdates);
+  }
+
+  public static Map<String, Collection<BeamFnApi.Metrics.User>> metricUpdatesToProto(
+      MetricUpdates metricUpdates) {
+
+    try {
+      return metricUpdatesToProtoOrThrow(metricUpdates);
+    } catch (ExecutionException exn) {
+      throw new RuntimeException("Impossible error converting metric updates to proto", exn);
```

Comment:
If it's actually impossible, this should be an `AssertionError`

It looks like you can use `getUnchecked` instead in `metricUpdatesToProtoOrThrow`, though, and then merge together these methods.